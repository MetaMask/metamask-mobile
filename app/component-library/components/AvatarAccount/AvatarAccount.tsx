/* eslint-disable react/prop-types */
import React from 'react';
import { Image } from 'react-native';
import JazzIcon from 'react-native-jazzicon';
import { toDataUrl } from '../../../util/blockies';
import Avatar from '../Avatar';
import stylesheet from './AvatarAccount.styles';
import { AvatarAccountProps, AvatarAccountType } from './AvatarAccount.types';

const AvatarAccount = ({
  type = AvatarAccountType.JazzIcon,
  accountAddress,
  size,
  style,
}: AvatarAccountProps) => (
  <Avatar size={size} style={style}>
    {
      {
        [AvatarAccountType.JazzIcon]: (
          <JazzIcon size={Number(size)} address={accountAddress} />
        ),
        [AvatarAccountType.Blockies]: (
          <Image
            source={{ uri: toDataUrl(accountAddress) }}
            style={stylesheet.imageStyle}
          />
        ),
      }[type]
    }
  </Avatar>
);

export default AvatarAccount;

export { AvatarAccount };
