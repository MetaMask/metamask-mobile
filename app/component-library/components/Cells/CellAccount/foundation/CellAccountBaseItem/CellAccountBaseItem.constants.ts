export const CELL_ACCOUNT_AVATAR_TEST_ID = 'cell-account-avatar';
export const CELL_ACCOUNT_TITLE_TEST_ID = 'cell-account-title';
export const CELL_ACCOUNT_SECONDARY_TEXT_TEST_ID =
  'cell-account-secondary-text';
export const CELL_ACCOUNT_TERTIARY_TEXT_TEST_ID = 'cell-account-tertiary-text';
export const CELL_ACCOUNT_TAG_LABEL_TEST_ID = 'cell-account-label';
