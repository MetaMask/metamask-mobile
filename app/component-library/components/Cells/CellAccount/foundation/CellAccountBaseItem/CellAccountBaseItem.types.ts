// Third party dependencies.
import { ViewProps } from 'react-native';

// External dependencies.
import { AvatarAccountType } from '../../../../Avatars/AvatarAccount';

/**
 * CellAccountBaseItem variants.
 */
export enum CellAccountItemType {
  Select = 'Select',
  Multiselect = 'Multiselect',
  Display = 'Display',
}

/**
 * Cell Account Display Item component props.
 */
export interface CellAccountBaseItemProps extends ViewProps {
  /**
   * Type of CellAccountItem
   */
  type?: CellAccountItemType;
  /**
   * An Ethereum wallet address to retrieve avatar.
   */
  avatarAccountAddress: string;
  /**
   * AvatarAccount variants.
   */
  avatarAccountType: AvatarAccountType;
  /**
   * Title of the Cell Account, 1 line truncation.
   */
  title: string;
  /**
   * Optional secondary text below the title, 1 line truncation.
   */
  secondaryText?: string;
  /**
   * Optional tertiary text below the secondaryText, 1 line truncation.
   */
  tertiaryText?: string;
  /**
   * Optional label (using Tag component) below the title/secondaryText/tertiaryText.
   */
  tagLabel?: string;
  /**
   * Optional accessory that can be inserted on the right of Cell Account.
   */
  children?: React.ReactNode;
}

/**
 * Style sheet input parameters.
 */
export type CellAccountBaseItemStyleSheetVars = Pick<
  CellAccountBaseItemProps,
  'style'
>;
