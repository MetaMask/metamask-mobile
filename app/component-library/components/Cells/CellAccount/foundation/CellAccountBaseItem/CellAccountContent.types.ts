// Third party dependencies.
import { ViewProps } from 'react-native';

// External dependencies.
import { AvatarAccountType } from '../../Avatars/AvatarAccount';

/**
 * Cell Account Content component props.
 */
export interface CellAccountContentProps extends ViewProps {
  /**
   * An Ethereum wallet address to retrieve avatar.
   */
  avatarAccountAddress: string;
  /**
   * AvatarAccount variants.
   */
  avatarAccountType: AvatarAccountType;
  /**
   * Title of the Cell Account, 1 line truncation.
   */
  title: string;
  /**
   * Optional secondary text below the title, 1 line truncation.
   */
  secondaryText?: string;
  /**
   * Optional tertiary text below the secondaryText, 1 line truncation.
   */
  tertiaryText?: string;
  /**
   * Optional label (using Tag component) below the title/secondaryText/tertiaryText.
   */
  tagLabel?: string;
  /**
   * Optional accessory that can be inserted on the right of Cell Account.
   */
  children?: React.ReactNode;
}

/**
 * Style sheet input parameters.
 */
export type CellAccountContentStyleSheetVars = Pick<
  CellAccountContentProps,
  'style'
>;
