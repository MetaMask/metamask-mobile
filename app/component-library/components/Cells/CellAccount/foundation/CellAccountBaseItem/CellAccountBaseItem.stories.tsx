/* eslint-disable no-console */

// Third party dependencies.
import React from 'react';
import { boolean, text } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react-native';

// External dependencies.
import { AvatarAccountType } from '../../../../Avatars/AvatarAccount';
import {
  TEST_ACCOUNT_ADDRESS,
  TEST_CELL_ACCOUNT_TITLE,
  TEST_CELL_ACCOUNT_SECONDARY_TEXT,
  TEST_CELL_ACCOUNT_TERTIARY_TEXT,
  TEST_TAG_LABEL_TEXT,
} from '../../CellAccount.constants';

// Internal dependencies.
import CellAccountBaseItem from './CellAccountBaseItem';

storiesOf('Component Library / CellAccountBaseItem', module).add(
  'Default',
  () => {
    const groupId = 'Props';
    const titleText = text('title', TEST_CELL_ACCOUNT_TITLE, groupId);
    const includeSecondaryText = boolean(
      'Includes secondaryText?',
      false,
      groupId,
    );
    const secondaryText = includeSecondaryText
      ? text('secondaryText', TEST_CELL_ACCOUNT_SECONDARY_TEXT, groupId)
      : '';
    const includeTertiaryText = boolean(
      'Includes tertiaryText?',
      false,
      groupId,
    );
    const tertiaryText = includeTertiaryText
      ? text('tertiaryText', TEST_CELL_ACCOUNT_TERTIARY_TEXT, groupId)
      : '';
    const includeTagLabel = boolean('Includes label?', false, groupId);
    const tagLabel = includeTagLabel
      ? text('label', TEST_TAG_LABEL_TEXT, groupId)
      : '';

    return (
      <CellAccountBaseItem
        avatarAccountAddress={TEST_ACCOUNT_ADDRESS}
        avatarAccountType={AvatarAccountType.JazzIcon}
        title={titleText}
        secondaryText={secondaryText}
        tertiaryText={tertiaryText}
        tagLabel={tagLabel}
      />
    );
  },
);
