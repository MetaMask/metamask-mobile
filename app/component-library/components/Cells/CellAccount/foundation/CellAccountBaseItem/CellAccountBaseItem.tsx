/* eslint-disable react/prop-types */

// Third library dependencies.
import React from 'react';
import { View } from 'react-native';

// External dependencies.
import { useStyles } from '../../../../../hooks';
import AvatarAccount from '../../../../Avatars/AvatarAccount';
import Text, { TextVariant } from '../../../../Text';
import { AvatarBaseSize } from '../../../../Avatars/AvatarBase';
import Tag from '../../../../Tags/Tag';

// Internal dependencies.
import {
  CELL_ACCOUNT_AVATAR_TEST_ID,
  CELL_ACCOUNT_TITLE_TEST_ID,
  CELL_ACCOUNT_SECONDARY_TEXT_TEST_ID,
  CELL_ACCOUNT_TERTIARY_TEXT_TEST_ID,
  CELL_ACCOUNT_TAG_LABEL_TEST_ID,
} from './CellAccountBaseItem.constants';
import styleSheet from './CellAccountBaseItem.styles';
import { CellAccountBaseItemProps } from './CellAccountBaseItem.types';

const CellAccountBaseItem = ({
  style,
  avatarAccountAddress,
  avatarAccountType,
  title,
  secondaryText,
  tertiaryText,
  tagLabel,
  children,
}: CellAccountBaseItemProps) => {
  const { styles } = useStyles(styleSheet, { style });

  return (
    <View style={styles.cellAccountBaseItem}>
      {/* DEV Note: Account Avatar should be replaced with Avatar with Badge whenever available */}
      <AvatarAccount
        type={avatarAccountType}
        accountAddress={avatarAccountAddress}
        size={AvatarBaseSize.Md}
        style={styles.avatarAccount}
        testID={CELL_ACCOUNT_AVATAR_TEST_ID}
      />
      <View style={styles.cellAccountBaseItemInfo}>
        <Text
          numberOfLines={1}
          variant={TextVariant.sHeadingSMRegular}
          testID={CELL_ACCOUNT_TITLE_TEST_ID}
        >
          {title}
        </Text>
        {!!secondaryText && (
          <Text
            numberOfLines={1}
            variant={TextVariant.sBodyMD}
            style={styles.secondaryText}
            testID={CELL_ACCOUNT_SECONDARY_TEXT_TEST_ID}
          >
            {secondaryText}
          </Text>
        )}
        {!!tertiaryText && (
          <Text
            numberOfLines={1}
            variant={TextVariant.sBodyMD}
            style={styles.tertiaryText}
            testID={CELL_ACCOUNT_TERTIARY_TEXT_TEST_ID}
          >
            {tertiaryText}
          </Text>
        )}
        {!!tagLabel && (
          <Tag
            label={tagLabel}
            style={styles.tagLabel}
            testID={CELL_ACCOUNT_TAG_LABEL_TEST_ID}
          />
        )}
      </View>
      {children && <View style={styles.optionalAccessory}>{children}</View>}
    </View>
  );
};

export default CellAccountBaseItem;
