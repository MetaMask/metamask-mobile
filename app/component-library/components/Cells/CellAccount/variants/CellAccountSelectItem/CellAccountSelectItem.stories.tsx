/* eslint-disable no-console */

// Third party dependencies.
import React from 'react';
import { Alert } from 'react-native';
import { boolean, text } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react-native';

// External dependencies.
import { AvatarAccountType } from '../../../../Avatars/AvatarAccount';
import {
  TEST_ACCOUNT_ADDRESS,
  TEST_CELL_ACCOUNT_TITLE,
  TEST_CELL_ACCOUNT_SECONDARY_TEXT,
  TEST_CELL_ACCOUNT_TERTIARY_TEXT,
  TEST_TAG_LABEL_TEXT,
} from '../../CellAccount.constants';
import { CellAccountItemType } from '../../foundation/CellAccountBaseItem/CellAccountBaseItem.types';

// Internal dependencies.
import CellAccountSelectItem from './CellAccountSelectItem';

storiesOf('Component Library / CellAccountSelectItem', module).add(
  'Default',
  () => {
    const groupId = 'Props';
    const titleText = text('title', TEST_CELL_ACCOUNT_TITLE, groupId);
    const includeSecondaryText = boolean(
      'Includes secondaryText?',
      false,
      groupId,
    );
    const secondaryText = includeSecondaryText
      ? text('secondaryText', TEST_CELL_ACCOUNT_SECONDARY_TEXT, groupId)
      : '';
    const includeTertiaryText = boolean(
      'Includes tertiaryText?',
      false,
      groupId,
    );
    const tertiaryText = includeTertiaryText
      ? text('tertiaryText', TEST_CELL_ACCOUNT_TERTIARY_TEXT, groupId)
      : '';
    const includeTagLabel = boolean('Includes label?', false, groupId);
    const tagLabel = includeTagLabel
      ? text('label', TEST_TAG_LABEL_TEXT, groupId)
      : '';
    const isSelected = boolean('isSelected?', false, groupId);

    return (
      <CellAccountSelectItem
        type={CellAccountItemType.Select}
        avatarAccountAddress={TEST_ACCOUNT_ADDRESS}
        avatarAccountType={AvatarAccountType.JazzIcon}
        title={titleText}
        secondaryText={secondaryText}
        tertiaryText={tertiaryText}
        tagLabel={tagLabel}
        isSelected={isSelected}
        onPress={() => Alert.alert('Pressed account Cell!')}
      />
    );
  },
);
