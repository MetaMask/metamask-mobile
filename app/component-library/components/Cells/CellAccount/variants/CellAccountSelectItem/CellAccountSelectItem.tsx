/* eslint-disable react/prop-types */

// Third library dependencies.
import React from 'react';

// External dependencies.
import { useStyles } from '../../../../../hooks';
import SelectItem from '../../../../Select/Select/SelectItem';
import CellAccountBaseItem from '../../foundation/CellAccountBaseItem';

// Internal dependencies.
import { CELL_ACCOUNT_SELECT_ITEM_TEST_ID } from './CellAccountSelectItem.constants';
import styleSheet from './CellAccountSelectItem.styles';
import { CellAccountSelectItemProps } from './CellAccountSelectItem.types';

const CellAccountSelectItem = ({
  style,
  avatarAccountAddress,
  avatarAccountType,
  title,
  secondaryText,
  tertiaryText,
  tagLabel,
  isSelected = false,
  children,
}: CellAccountSelectItemProps) => {
  const { styles } = useStyles(styleSheet, { style });

  return (
    <SelectItem
      isSelected={isSelected}
      style={styles.base}
      testID={CELL_ACCOUNT_SELECT_ITEM_TEST_ID}
    >
      <CellAccountBaseItem
        avatarAccountAddress={avatarAccountAddress}
        avatarAccountType={avatarAccountType}
        title={title}
        secondaryText={secondaryText}
        tertiaryText={tertiaryText}
        tagLabel={tagLabel}
      >
        {children}
      </CellAccountBaseItem>
    </SelectItem>
  );
};

export default CellAccountSelectItem;
