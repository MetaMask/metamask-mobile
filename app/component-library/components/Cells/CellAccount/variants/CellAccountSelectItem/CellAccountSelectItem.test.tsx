// Third party dependencies.
import React from 'react';
import { shallow } from 'enzyme';

// External dependencies.
import { AvatarAccountType } from '../../../../Avatars/AvatarAccount';
import {
  TEST_ACCOUNT_ADDRESS,
  TEST_CELL_ACCOUNT_TITLE,
} from '../../CellAccount.constants';
import { CellAccountItemType } from '../../foundation/CellAccountBaseItem/CellAccountBaseItem.types';

// Internal dependencies.
import CellAccountSelectItem from './CellAccountSelectItem';
import { CELL_ACCOUNT_SELECT_ITEM_TEST_ID } from './CellAccountSelectItem.constants';

describe('CellAccountSelectItem - Snapshot', () => {
  it('should render default settings correctly', () => {
    const wrapper = shallow(
      <CellAccountSelectItem
        type={CellAccountItemType.Select}
        avatarAccountAddress={TEST_ACCOUNT_ADDRESS}
        avatarAccountType={AvatarAccountType.JazzIcon}
        title={TEST_CELL_ACCOUNT_TITLE}
        onPress={jest.fn}
      />,
    );
    expect(wrapper).toMatchSnapshot();
  });
  it('should render the proper selected state', () => {
    const wrapper = shallow(
      <CellAccountSelectItem
        type={CellAccountItemType.Select}
        avatarAccountAddress={TEST_ACCOUNT_ADDRESS}
        avatarAccountType={AvatarAccountType.JazzIcon}
        title={TEST_CELL_ACCOUNT_TITLE}
        isSelected
        onPress={jest.fn}
      />,
    );
    expect(wrapper).toMatchSnapshot();
  });
});

describe('CellAccountSelectItem', () => {
  it('should render singleSelect', () => {
    const wrapper = shallow(
      <CellAccountSelectItem
        type={CellAccountItemType.Select}
        avatarAccountAddress={TEST_ACCOUNT_ADDRESS}
        avatarAccountType={AvatarAccountType.JazzIcon}
        title={TEST_CELL_ACCOUNT_TITLE}
        onPress={jest.fn}
      />,
    );
    const singleSelectComponent = wrapper.findWhere(
      (node) => node.prop('testID') === CELL_ACCOUNT_SELECT_ITEM_TEST_ID,
    );
    expect(singleSelectComponent.exists()).toBe(true);
  });
});
