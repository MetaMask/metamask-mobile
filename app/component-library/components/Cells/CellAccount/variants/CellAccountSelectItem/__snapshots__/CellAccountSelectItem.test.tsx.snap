// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CellAccountSelectItem - Snapshot should render default settings correctly 1`] = `
<SelectItem
  isSelected={false}
  style={Object {}}
>
  <CellAccountBaseItem
    avatarAccountAddress="0x2990079bcdEe240329a520d2444386FC119da21a"
    avatarAccountType="JazzIcon"
    title="Orangefox.eth"
  />
</SelectItem>
`;

exports[`CellAccountSelectItem - Snapshot should render the proper selected state 1`] = `
<SelectItem
  isSelected={true}
  style={Object {}}
>
  <CellAccountBaseItem
    avatarAccountAddress="0x2990079bcdEe240329a520d2444386FC119da21a"
    avatarAccountType="JazzIcon"
    title="Orangefox.eth"
  />
</SelectItem>
`;
